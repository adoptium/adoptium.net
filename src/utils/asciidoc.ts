import Asciidoctor from "@asciidoctor/core";
import fs from "fs";
import path from "path";
import { decode } from "html-entities";

interface AsciidoctorDocument {
  getSourceLocation(): { getPath(): string };
}

interface AsciidoctorReader {
  pushInclude(
    content: string,
    file: string,
    target: string,
    line: number,
    attributes: Record<string, unknown>
  ): void;
}

interface AsciidoctorAttributes {
  [key: string]: unknown;
}

// Initialize Asciidoctor processor
const asciidoctor = Asciidoctor();

/**
 * Custom Asciidoctor include processor that resolves partials
 */
class CustomIncludeProcessor {
  private basePath: string;

  constructor(basePath: string) {
    this.basePath = basePath;
  }

  process(
    doc: AsciidoctorDocument,
    reader: AsciidoctorReader,
    target: string,
    attributes: AsciidoctorAttributes
  ) {
    // Get current file directory to resolve relative includes correctly
    const currentDir = path.dirname(doc.getSourceLocation().getPath());

    let includeFile = target;

    // If target is a relative path, resolve it from current file directory
    if (!path.isAbsolute(target) && !target.startsWith("/")) {
      includeFile = path.resolve(currentDir, target);
    }

    // If not found, try from the content base path
    if (!fs.existsSync(includeFile)) {
      includeFile = path.join(this.basePath, target);
    }

    if (fs.existsSync(includeFile)) {
      const content = fs.readFileSync(includeFile, "utf8");
      return reader.pushInclude(content, includeFile, target, 1, attributes);
    } else {
      console.warn(`Include file not found: ${includeFile}`);
      return reader.pushInclude(
        `[WARNING] Include file not found: ${target}`,
        target,
        target,
        1,
        attributes
      );
    }
  }
}

/**
 * Process AsciiDoc content with proper include resolution
 */
export function processAsciiDoc(
  filePath: string,
  content: string,
  options: Record<string, unknown> = {}
) {
  const basePath = path.dirname(filePath);
  const registry = asciidoctor.Extensions.create();

  // Create an instance of our custom processor
  const customProcessor = new CustomIncludeProcessor(basePath);

  // Try using a simple approach first - just skip extension registration
  // and see if the document still processes correctly
  const skipExtensions = true;

  if (!skipExtensions) {
    // Register include processor using a factory function
    const IncludeProcessor = function () {};
    IncludeProcessor.prototype.handles = function () {
      return true;
    };
    IncludeProcessor.prototype.process = function (
      doc: AsciidoctorDocument,
      reader: AsciidoctorReader,
      target: string,
      attrs: AsciidoctorAttributes
    ) {
      return customProcessor.process(doc, reader, target, attrs);
    };

    registry.includeProcessor(IncludeProcessor);
  }

  // Set default options with proper typing
  const defaultOptions: Record<string, unknown> = {
    safe: "server",
    base_dir: basePath,
    attributes: {
      "source-highlighter": "highlight.js",
      icons: "font",
      sectanchors: "",
      idprefix: "",
      idseparator: "-",
    },
  };

  // Add extension registry only if we're not skipping extensions
  if (!skipExtensions) {
    defaultOptions.extension_registry = registry;
  }

  // Merge options
  const mergedOptions = { ...defaultOptions, ...options };

  // Convert AsciiDoc to HTML - use the processor without extension registration
  // if we're having issues with extensions
  try {
    let html = asciidoctor.convert(content, mergedOptions) as string;
    // Post-process image src paths: replace src="filename.png" with src="/docs/filename.png" if not already absolute or external
    html = html.replace(
      /<img([^>]+)src=["'](?![a-z]+:|\/)([^"'>]+)["']/gi,
      '<img$1src="/docs/$2"'
    );
    return html;
  } catch (error) {
    console.error("Error converting AsciiDoc content:", error);

    // Try without extension registry as fallback
    // Create a new options object without the extension_registry property using destructuring
    // Use object destructuring with "rest" syntax to remove the property
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { extension_registry, ...fallbackOptions } = defaultOptions;
    let html = asciidoctor.convert(content, fallbackOptions) as string;
    html = html.replace(
      /<img([^>]+)src=["'](?![a-z]+:|\/)([^"'>]+)["']/gi,
      '<img$1src="/docs/$2"'
    );
    return html;
  }
}

/**
 * Decode HTML entities in a string
 */
export function decodeHtmlEntities(text: string): string {
  return decode(text);
}

/**
 * Extract metadata from AsciiDoc document
 */
export function extractMetadata(
  filePath: string,
  content: string
): {
  title: string;
  description: string;
  authors: string[];
  attributes: Record<string, unknown>;
} {
  const doc = asciidoctor.load(content, { safe: "server" });

  // Get the title and ensure it's a string
  const docTitle = doc.getDocumentTitle();
  let title =
    typeof docTitle === "string"
      ? docTitle
      : docTitle?.toString() || path.basename(filePath, path.extname(filePath));

  // Decode HTML entities in the title
  title = decodeHtmlEntities(title);

  // Get and decode description if it exists
  const description = doc.hasAttribute("description")
    ? decodeHtmlEntities(String(doc.getAttribute("description")))
    : "";

  return {
    title,
    description,
    authors: doc.hasAttribute("page-authors")
      ? String(doc.getAttribute("page-authors"))
          .split(",")
          .map((author: string) => author.trim())
      : [],
    attributes: doc.getAttributes() as Record<string, unknown>,
  };
}

/**
 * Check if a file exists
 */
export function fileExists(filePath: string): boolean {
  try {
    return fs.existsSync(filePath);
  } catch {
    // Silently handle any errors and return false
    return false;
  }
}

/**
 * Get all available language variants for a given base path
 */
export function getLanguageVariants(
  basePath: string,
  fileName: string
): string[] {
  const dir = path.dirname(basePath);
  const baseFileName = path.basename(fileName, ".adoc");

  // Get all adoc files in the directory
  const files = fs
    .readdirSync(dir)
    .filter((file) => file.startsWith(baseFileName) && file.endsWith(".adoc"));

  // Extract locales from filenames
  return files.map((file) => {
    // Format: index.de.adoc -> de
    const match = file.match(
      new RegExp(`${baseFileName}(?:\\.([\\w-]+))?\\.adoc$`)
    );
    return match && match[1] ? match[1] : "en";
  });
}
