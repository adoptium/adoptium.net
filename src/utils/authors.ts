import AuthorData from "@/data/authors.json";

export interface Author {
  name: string;
  summary?: string;
  twitter?: string;
  github?: string;
  linkedin?: string;
}

export function getFormattedAuthorData(authorId: string): Author {
  const rawAuthorData = AuthorData[authorId as keyof typeof AuthorData];

  if (!rawAuthorData) {
    return {
      name: "Unknown Author",
    };
  }

  // Convert the raw author data to our Author interface
  return {
    name: rawAuthorData.name || "Unknown Author",
    // Convert null values to undefined
    summary: rawAuthorData.summary || undefined,
    twitter: rawAuthorData.twitter || undefined,
    github: rawAuthorData.github || undefined,
    linkedin: rawAuthorData.linkedin || undefined,
  };
}
