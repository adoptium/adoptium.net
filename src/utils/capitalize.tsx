// Format text with a capital letter
export function capitalize(text: string) {
  switch (text) {
    case "alpine-linux":
      return "Alpine Linux"
    case "alpine_linux":
      return "Alpine Linux"
    case "redhat":
      return "Red Hat"
    case "ibm":
      return "IBM"
    case "mac":
      return "macOS"
    case "semeru":
      return "Semeru Runtimes"
    default:
      return text.charAt(0).toUpperCase() + text.slice(1)
  }
}
