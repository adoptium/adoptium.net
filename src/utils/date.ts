export function formatDate(
  date: string = "",
  locale: string = "en-US"
): string {
  const targetDate = date ? new Date(date) : new Date();
  return targetDate.toLocaleDateString(locale, {
    month: "long",
    day: "2-digit",
    year: "numeric",
  });
}
