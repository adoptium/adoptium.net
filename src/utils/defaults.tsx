// This file is used to store defaults which can be accessed by other Javascript/Typescript functions

export const packageTypes = [
  {
    name: "JDK",
    value: "jdk",
  },
  {
    name: "JRE",
    value: "jre",
  },
]

const defaultPackageType = "jdk"
const defaultArchitecture = "x64"

export { defaultPackageType, defaultArchitecture }
