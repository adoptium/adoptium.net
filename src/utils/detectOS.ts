export enum UserOS {
  MAC = "MAC",
  WIN = "WIN",
  UNIX = "UNIX",
  LINUX = "LINUX",
  MOBILE = "MOBILE",
  UNKNOWN = "UNKNOWN",
}

export function detectOS(): UserOS {
  let OS = UserOS.UNKNOWN
  if (typeof window !== "undefined") {
    const appVersion = navigator.appVersion;
    if (appVersion.includes("Win")) {
      OS = UserOS.WIN
    } else if (appVersion.includes("Mac")) {
      OS = UserOS.MAC
    } else if (appVersion.includes("X11")) {
      OS = UserOS.UNIX
    } else if (appVersion.includes("Linux")) {
      OS = UserOS.LINUX
    } else if (appVersion.includes("Mobi")) {
      OS = UserOS.MOBILE
    }
  }
  return OS
}
