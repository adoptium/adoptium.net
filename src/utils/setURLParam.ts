export function setURLParam(param: string, value: string | number | boolean | undefined) {
  const currentURLWithParams = window.location.href

  const params: Record<string, string> = {}

  // parse existing query parameters (if any)
  if (currentURLWithParams.indexOf("?") > 0) {
    currentURLWithParams
      .substring(currentURLWithParams.indexOf("?") + 1)
      .split("&")
      .forEach(item => {
        params[item.split("=")[0]] = item.split("=")[1]
      })
  }
  if (param) {
    // set the new value for the given param or unset if no value provided
    if (value === undefined) delete params[param]
    else params[param] = String(value) // Convert value to string
  }

  // rebuild and set the location
  const queryString = Object.keys(params)
    .map(key => {
      return encodeURIComponent(key) + "=" + encodeURIComponent(params[key])
    })
    .join("&")

  const currentURL =
    window.location.protocol +
    "//" +
    window.location.host +
    window.location.pathname +
    `?${queryString}`
  window.history.replaceState({ path: currentURL }, "", currentURL)
}
