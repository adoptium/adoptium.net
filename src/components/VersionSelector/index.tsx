"use client"

import React from "react"
import { setURLParam } from "@/utils/setURLParam"

interface VersionSelectorProps {
  activeVersionTab: number | string
  setActiveVersionTab: (tab: number | string) => void
  versions: Array<{ name: string, value: number }>
  updateVersion: (version: number | string) => void
  defaultVersion: string
  updateOS?: (os: string) => void
  updateArch?: (arch: string) => void
}

const VersionSelector: React.FC<VersionSelectorProps> = ({
  activeVersionTab,
  setActiveVersionTab,
  versions,
  updateVersion,
  updateOS,
  updateArch,
}) => {
  // Handle tab selection with proper URL and state updates
  const handleTabClick = (newTab: number | string) => {
    // Notify parent of tab change
    setActiveVersionTab(newTab)

    // If All Versions tab is clicked (tab 1)
    if (newTab === 1) {
      // For All Versions tab, we'll preserve the current version
      // Don't need to update version in URL, parent component will handle this

      // Reset OS and arch to "any"
      setURLParam("os", "any")
      if (updateOS) updateOS("any")
      setURLParam("arch", "any")
      if (updateArch) updateArch("any")
    }
    // If specific version tab is clicked
    else {
      // Update version parameter
      setURLParam("version", newTab)
      updateVersion(newTab)

      // Reset OS and arch to "any"
      setURLParam("os", "any")
      if (updateOS) updateOS("any")
      setURLParam("arch", "any")
      if (updateArch) updateArch("any")
    }
  }

  return (
    <div className="w-full max-w-[1264px] mx-auto ">
      <div className="relative min-w-full w-full overflow-auto md:overflow-visible">
        <span className="h-[1px] w-full bg-[#3E3355] inline-block absolute bottom-0 z-[-1]"></span>
        <div className="flex space-x-10 whitespace-nowrap lg:justify-center py-3">
          <button onClick={() => handleTabClick(1)}>
            <span
              className={`py-3 w-full text-base font-normal leading-6 
            outline-none cursor-pointer transition-all duration-200 ease-in-out ${activeVersionTab === 1 ? "border-primary border-b-[2px] text-white" : "text-[#8a809e] border-transparent border-b"}`}
            >
              All Versions
            </span>
          </button>
          {versions.map((version, index) => (
            <button
              key={index}
              onClick={() => handleTabClick(version.value)}
            >
              <span
                className={`py-3 w-full text-base font-normal leading-6
            outline-none cursor-pointer transition-all duration-200 ease-in-out ${activeVersionTab === version.value ? "border-primary border-b-[2px] text-white" : "text-[#8a809e] border-transparent border-b"}`}
              >
                JDK {version.name}
              </span>
            </button>
          ))}
        </div>
      </div>
    </div>
  )
}

export default VersionSelector