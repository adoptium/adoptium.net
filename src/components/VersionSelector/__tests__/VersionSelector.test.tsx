import React from "react"
import { render, fireEvent } from "@testing-library/react"
import { beforeEach, describe, expect, it, vi } from "vitest"
import { axe } from "vitest-axe"
import VersionSelector from "../index"
import { setURLParam } from "@/utils/setURLParam"

// Mock the setURLParam utility
vi.mock("@/utils/setURLParam", () => ({
  setURLParam: vi.fn(),
}))

describe("VersionSelector component", () => {
  const mockProps = {
    activeVersionTab: 1,
    setActiveVersionTab: vi.fn(),
    versions: [
      {
        name: "8",
        value: 8
      },
      {
        name: "11",
        value: 11
      },
      {
        name: "17",
        value: 17
      },
    ],
    updateVersion: vi.fn(),
    defaultVersion: "latest",
    updateOS: vi.fn(),
    updateArch: vi.fn(),
  }

  beforeEach(() => {
    vi.clearAllMocks()
  })

  it("renders correctly", () => {
    const { container } = render(<VersionSelector {...mockProps} />)
    expect(container).toMatchSnapshot()
  })

  it("renders 'All Versions' button and version buttons", () => {
    const { container } = render(<VersionSelector {...mockProps} />)

    // Get all buttons
    const buttons = container.querySelectorAll('button');

    // Check for button content by getting the text content
    expect(buttons[0].textContent).toContain("All Versions")
    expect(buttons[1].textContent).toContain("JDK 8")
    expect(buttons[2].textContent).toContain("JDK 11")
    expect(buttons[3].textContent).toContain("JDK 17")
  })

  it("highlights the active version tab", () => {
    const activeProps = {
      ...mockProps,
      activeVersionTab: 11,
    }

    const { container } = render(<VersionSelector {...activeProps} />)

    // Use container.querySelector to find the elements by their exact position
    // Find the first button (All Versions) and the third button (JDK 11)
    const buttons = container.querySelectorAll('button');
    const allVersionsSpan = buttons[0].querySelector('span');
    const jdk11Span = buttons[2].querySelector('span');

    // Check correct classes are applied
    expect(allVersionsSpan).toHaveClass("text-[#8a809e]")
    expect(allVersionsSpan).toHaveClass("border-transparent")

    expect(jdk11Span).toHaveClass("text-white")
    expect(jdk11Span).toHaveClass("border-primary")
    expect(jdk11Span).toHaveClass("border-b-[2px]")
  })

  it("calls setActiveTabVersion with specific version when a version button is clicked", () => {
    const { container } = render(<VersionSelector {...mockProps} />)

    // Get the JDK 11 button directly by its position (third button)
    const buttons = container.querySelectorAll('button');
    const jdk11Button = buttons[2]; // Index 2 corresponds to the third button (JDK 11)

    // Click on JDK 11 button
    fireEvent.click(jdk11Button)

    // Verify the callbacks
    expect(setURLParam).toHaveBeenCalledWith("version", 11)
    expect(mockProps.updateVersion).toHaveBeenCalledWith(11)
    expect(mockProps.setActiveVersionTab).toHaveBeenCalledWith(11)

    // Verify OS and arch are reset to "any"
    expect(setURLParam).toHaveBeenCalledWith("os", "any")
    expect(mockProps.updateOS).toHaveBeenCalledWith("any")
    expect(setURLParam).toHaveBeenCalledWith("arch", "any")
    expect(mockProps.updateArch).toHaveBeenCalledWith("any")
  })

  it("has no accessibility violations", async () => {
    const { container } = render(<VersionSelector {...mockProps} />)
    const results = await axe(container)

    // Check for violations
    expect(results.violations.length).toBe(0)
  })
})