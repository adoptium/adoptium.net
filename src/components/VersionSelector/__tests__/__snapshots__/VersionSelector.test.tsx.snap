// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VersionSelector component > renders correctly 1`] = `
<div>
  <div
    class="w-full max-w-[1264px] mx-auto "
  >
    <div
      class="relative min-w-full w-full overflow-auto md:overflow-visible"
    >
      <span
        class="h-[1px] w-full bg-[#3E3355] inline-block absolute bottom-0 z-[-1]"
      />
      <div
        class="flex space-x-10 whitespace-nowrap lg:justify-center py-3"
      >
        <button>
          <span
            class="py-3 w-full text-base font-normal leading-6 
            outline-none cursor-pointer transition-all duration-200 ease-in-out border-primary border-b-[2px] text-white"
          >
            All Versions
          </span>
        </button>
        <button>
          <span
            class="py-3 w-full text-base font-normal leading-6
            outline-none cursor-pointer transition-all duration-200 ease-in-out text-[#8a809e] border-transparent border-b"
          >
            JDK 
            8
          </span>
        </button>
        <button>
          <span
            class="py-3 w-full text-base font-normal leading-6
            outline-none cursor-pointer transition-all duration-200 ease-in-out text-[#8a809e] border-transparent border-b"
          >
            JDK 
            11
          </span>
        </button>
        <button>
          <span
            class="py-3 w-full text-base font-normal leading-6
            outline-none cursor-pointer transition-all duration-200 ease-in-out text-[#8a809e] border-transparent border-b"
          >
            JDK 
            17
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
